Runtime semantics difference under better-fors

## Compiler version

3.8

Workaround: Run with `-source:3.7`

## Minimized code

```scala
def f(x: Int): (Int, Int) = (1, 0)

val result: Iterable[(Int, Int)] =
  for
    (k, v)   <- Map(1 -> 1, 2 -> 1, 3 -> 1)
    x        = k + v
    (a, b)   = f(x)
  yield (a, b)

@main def Test() = 
  assert(result.size == 3, "Expected 3 elements, got ${result.size}")
```

## Output

```scala
> scala-cli run -S 3.7 bug.scala
Compiling project (Scala 3.7.4, JVM (21))
Compiled project (Scala 3.7.4, JVM (21))
> scala-cli run -S 3.nightly bug.scala
Compiling project (Scala 3.8.3-RC1-bin-20260123-506b8a0-NIGHTLY, JVM (21))
Compiled project (Scala 3.8.3-RC1-bin-20260123-506b8a0-NIGHTLY, JVM (21))
Exception in thread "main" java.lang.AssertionError: assertion failed: Expected 3 elements, got 1
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:10)
        at bug$package$.Test(bug.scala:11)
        at Test.main(bug.scala:10)
```

## Expectation
Should produce the same results as version without better-fors (using Scala 3.7, or 3.8 with -source:3.7)

--- a/compiler/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/compiler/src/dotty/tools/dotc/ast/Desugar.scala
@@ -64,9 +64,16 @@ object desugar {
   val PolyFunctionApply: Property.Key[Unit] = Property.StickyKey()
 
   /** An attachment key to indicate that an Apply is created as a last `map`
-   *  scall in a for-comprehension.
+   *  call in a for-comprehension.
    */
   val TrailingForMap: Property.Key[Unit] = Property.StickyKey()
+
+  /** An attachment key to indicate that an Apply (`map` or `flatMap`)
+   *  in a for-comprehension has a nested tupling operation that no longer
+   *  generates an extra `map` call. That call would select the other overload
+   *  of `Map#map` and change the result to `List`. `DropForMap` warns about it.
+   */
+  val TuplingMigrationForMap: Property.Key[Unit] = Property.StickyKey()
 
   val WasTypedInfix: Property.Key[Unit] = Property.StickyKey()
 
@@ -2286,7 +2293,9 @@ object desugar {
             val selectName =
               if suffix.exists(_.isInstanceOf[GenFrom]) then flatMapName
               else mapName
-            Apply(rhsSelect(gen, selectName), makeLambda(gen, cont))
+            val app = Apply(rhsSelect(gen, selectName), makeLambda(gen, cont))
+            if valeqs.lengthIs > 1 then app.withAttachment(TuplingMigrationForMap, ())
+            else app
           else
             val (pats, rhss) = valeqs.map { case GenAlias(pat, rhs) => (pat, rhs) }.unzip
             val (defpat0, id0) = makeIdPat(gen.pat)

--- a/compiler/src/dotty/tools/dotc/transform/localopt/DropForMap.scala
+++ b/compiler/src/dotty/tools/dotc/transform/localopt/DropForMap.scala
@@ -1,7 +1,7 @@
 package dotty.tools.dotc
 package transform.localopt
 
-import dotty.tools.dotc.ast.desugar.TrailingForMap
+import dotty.tools.dotc.ast.desugar.{TrailingForMap, TuplingMigrationForMap}
 import dotty.tools.dotc.ast.tpd.*
 import dotty.tools.dotc.core.Contexts.*
 import dotty.tools.dotc.core.Decorators.*
@@ -30,7 +30,23 @@ class DropForMap extends MiniPhase:
   /** r.map(x => x)(using y) --> r
    *       ^ TrailingForMap
    */
-  override def transformApply(tree: Apply)(using Context): Tree = tree match
+  override def transformApply(tree: Apply)(using Context): Tree =
+    tree.removeAttachment(TuplingMigrationForMap).foreach: _ =>
+      def stripApplies(tree: Tree): Tree = tree match
+        case Apply(fun, _)     => stripApplies(fun)
+        case TypeApply(fun, _) => stripApplies(fun)
+        case tree => tree
+      stripApplies(tree) match
+      case Select(coll, _) =>
+        val m = coll.tpe.nonPrivateMember(nme.map)
+        if m.isOverloaded then
+          val alts = m.alternatives
+          val head = alts.head.info.finalResultType
+          if !alts.tail.forall(_.info.finalResultType =:= head) then
+            val msg = em"For comprehension with multiple val assignments may change result type in 3.8"
+            report.warning(msg, tree.srcPos)
+      case _ =>
+    tree match
     case Unmapped(f0, sym, args) =>
       val f =
         if sym.is(Extension) then args.head

--- a/tests/neg/i25077.check
+++ b/tests/neg/i25077.check
@@ -1,0 +1,8 @@
+-- Error: tests/neg/i25077.scala:6:2 -----------------------------------------------------------------------------------
+ 6 |  for // error behavior change in 3.8
+   |  ^
+   |  For comprehension with multiple val assignments may change result type in 3.8
+   |
+ 7 |    (k, v) <- Map(1 -> 1, 2 -> 1, 3 -> 1)
+ 8 |...
+11 |  yield (a+y, b+z)

--- a/tests/neg/i25077.scala
+++ b/tests/neg/i25077.scala
@@ -1,0 +1,11 @@
+//> using options "-Wconf:msg=multiple val:e"
+
+def f(x: Int): (Int, Int) = (1, x)
+
+val result: Iterable[(Int, Int)] =
+  for // error behavior change in 3.8
+    (k, v) <- Map(1 -> 1, 2 -> 1, 3 -> 1)
+    x      = k + v
+    (a, b) = f(x)
+    (y, z) <- Map(42 -> 27)
+  yield (a+y, b+z)

--- a/tests/run/i25077.scala
+++ b/tests/run/i25077.scala
@@ -1,0 +1,70 @@
+def f(x: Int): (Int, Int) = (1, x)
+
+val result: Iterable[(Int, Int)] =
+  for
+    (k, v) <- Map(1 -> 1, 2 -> 1, 3 -> 1)
+    x      = k + v
+    (a, b) = f(x) // warn previously map would have changed result type
+    (y, z) <- Map(42 -> 27)
+  yield (a+y, b+z)
+
+@main def Test() =
+  assert(result.size == 1, s"Expected 1 only, got ${result.size}")
+  assert(result.head == (43, 31), result.head)
+
+/*
+3.8
+
+    val result: Iterable[Tuple2[Int, Int]] =
+      Map.apply[Int, Int](
+        [ArrowAssoc[Int](1).->[Int](1),ArrowAssoc[Int](2).->[Int](1),
+          ArrowAssoc[Int](3).->[Int](1) : (Int, Int)]*
+      ).map[Int, Int]((x$1: (Int, Int)) =>
+        x$1:(Int, Int) @RuntimeChecked match
+          {
+            case Tuple2.unapply[Int, Int](k @ _, v @ _) =>
+              val x: Int = k + v
+              val $2$: ((Int, Int), Int, Int) =
+                f(x):(Int, Int) @RuntimeChecked match
+                  {
+                    case $1$ @ Tuple2.unapply[Int, Int](a @ _, b @ _) =>
+                      Tuple3.apply[(Int, Int), Int, Int]($1$, a, b)
+                  }
+              val $1$: (Int, Int) = $2$._1
+              val a: Int = $2$._2
+              val b: Int = $2$._3
+              Tuple2.apply[Int, Int](a, b)
+          }
+      )
+3.7
+    val result: Iterable[Tuple2[Int, Int]] =
+      Map.apply[Int, Int](
+        [ArrowAssoc[Int](1).->[Int](1),ArrowAssoc[Int](2).->[Int](1),
+          ArrowAssoc[Int](3).->[Int](1) : (Int, Int)]*
+      ).map[((Int, Int), Int, (Int, Int))]((x$1: (Int, Int)) =>
+        x$1:(Int, Int) @unchecked match
+          {
+            case $1$ @ Tuple2.unapply[Int, Int](k @ _, v @ _) =>
+              val x: Int = k + v
+              val $3$: ((Int, Int), Int, Int) =
+                f(x):(Int, Int) @unchecked match
+                  {
+                    case $2$ @ Tuple2.unapply[Int, Int](a @ _, b @ _) =>
+                      Tuple3.apply[(Int, Int), Int, Int]($2$, a, b)
+                  }
+              val $2$: (Int, Int) = $3$._1
+              val a: Int = $3$._2
+              val b: Int = $3$._3
+              Tuple3.apply[(Int, Int), Int, (Int, Int)]($1$, x, $2$)
+          }
+      ).map[(Int, Int)]((x$1: ((Int, Int), Int, (Int, Int))) =>
+        x$1:(x$1 : ((Int, Int), Int, (Int, Int))) @unchecked match
+          {
+            case
+              Tuple3.unapply[(Int, Int), Int, (Int, Int)](
+                Tuple2.unapply[Int, Int](k @ _, v @ _), x @ _,
+                Tuple2.unapply[Int, Int](a @ _, b @ _))
+             => Tuple2.apply[Int, Int](a, b)
+          }
+      )
+*/

--- a/tests/run/i25089.check
+++ b/tests/run/i25089.check
@@ -1,0 +1,2 @@
+Printed for both versions
+Printed only with 3.7.4

--- a/tests/run/i25089.scala
+++ b/tests/run/i25089.scala
@@ -1,0 +1,9 @@
+def runAsRight[T](execute: => T): Either[String, T] = Right(execute)
+
+@main def Test =
+  for
+    _ <- runAsRight(
+      println("Printed for both versions"),
+    )
+    _ = println(s"Printed only with 3.7.4")
+  yield ()

